h2/src/main/org/h2/expression/function/Function.java

```diff
@@ -1534,11 +1547,21 @@ private static Value regexpReplace(SessionLocal session, String input, String re
         }
         boolean isInPostgreSqlMode = Mode.ModeEnum.PostgreSQL.equals(mode.getEnum());
         int flags = makeRegexpFlags(regexpMode, isInPostgreSqlMode);
+        if (isInPostgreSqlMode && ( regexpMode == null || regexpMode.isEmpty() || !regexpMode.contains("g"))) {
+            occurrence = 1;
+        }
         try {
-            Matcher matcher = Pattern.compile(regexp, flags).matcher(input);
-            return ValueVarchar.get(isInPostgreSqlMode && (regexpMode == null || regexpMode.indexOf('g') < 0) ?
-                    matcher.replaceFirst(replacement) : matcher.replaceAll(replacement),
-                    session);
+            Matcher matcher = Pattern.compile(regexp, flags).matcher(input).region(position - 1, input.length());
+            if (occurrence == 0) {
+                return ValueVarchar.get(matcher.replaceAll(replacement), session);
+            } else {
+                for(int i = 0; i < occurrence; i++) {
+                    matcher.find();
+                }
+                int startIndex = matcher.start();
+                return ValueVarchar.get(input.substring(0, startIndex) + input.substring(startIndex).replaceFirst(regexp,
+                        replacement), session);
+            }
```

Comment:
Please, don't compile a new `Pattern` for the same regular expression here. Try to use an existing `Matcher` directly.